File: src/main/java/com/baidu/algorithm/bitmanipulation/_260_Single_Number_III.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        xor_val = 0
        for x in nums:
            xor_val ^= x

        bit_pos = 1
        while not (xor_val & bit_pos):
            bit_pos <<= 1

        v1, v2 = 0, 0
        for x in nums:
            if x & bit_pos:
                v1 ^= x
            else:
                v2 ^= x

        return [v1, v2]

File: src/main/java/com/baidu/algorithm/bitmanipulation/_260_Single_Number_III.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.bitmanipulation;

/**
 * _260_Single_Number_III
 *
 * @author xuhaoran01
 */
public class _260_Single_Number_III {
    public int[] singleNumber(int[] nums) {
        if(nums == null || nums.length < 2) {
            return new int[0];
        }

        int xorRes = 0;
        for(int num: nums) {
            xorRes ^= num;
        }

        if(xorRes == 0) {
            return new int[0];
        }

        int shift = 1;
        while((shift & xorRes) == 0) {
            shift <<= 1;
        }

        int first = 0, second = 0;
        for(int num: nums) {
            if((num & shift) == 0) {
                first ^= num;
            }
            else {
                second ^= num;
            }
        }

        return new int[]{first, second};
    }

}
